x/ccv/provider/ibc_middleware.go

```diff
@@ -0,0 +1,239 @@
+func (im IBCMiddleware) OnChanOpenTry(
+	ctx sdk.Context,
+	order channeltypes.Order,
+	connectionHops []string,
+	portID,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	counterpartyVersion string,
+) (string, error) {
+	// call underlying app's OnChanOpenTry callback with the appVersion
+	return im.app.OnChanOpenTry(ctx, order, connectionHops, portID, channelID, chanCap, counterparty, counterpartyVersion)
+}
+
+// OnChanOpenAck implements the IBCMiddleware interface
+func (im IBCMiddleware) OnChanOpenAck(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+	counterpartyChannelID string,
+	counterpartyVersion string,
+) error {
+	// call underlying app's OnChanOpenAck callback with the counterparty app version.
+	return im.app.OnChanOpenAck(ctx, portID, channelID, counterpartyChannelID, counterpartyVersion)
+}
+
+// OnChanOpenConfirm implements the IBCMiddleware interface
+func (im IBCMiddleware) OnChanOpenConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// call underlying app's OnChanOpenConfirm callback.
+	return im.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+// OnChanCloseInit implements the IBCMiddleware interface
+func (im IBCMiddleware) OnChanCloseInit(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// call underlying app's OnChanCloseInit callback.
+	return im.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+// OnChanCloseConfirm implements the IBCMiddleware interface
+func (im IBCMiddleware) OnChanCloseConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	return im.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnRecvPacket executes the IBC transfer. In case of success,
+// it verifies if the packet sender is a consumer chain
+// and if the received IBC coin is whitelisted. In such instances,
+// it appends the coin to the consumer's chain allocation record
+func (im IBCMiddleware) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	// executes the IBC transfer OnRecv logic
+	ack := im.app.OnRecvPacket(ctx, packet, relayer)
+
+	// execute the middleware logic only if the sender is a consumer chain
+	consumerID, err := im.keeper.IdentifyConsumerChainIDFromIBCPacket(ctx, packet)
+	if err != nil {
+		return ack
+	}
+
+	// Note that inside the above if condition statement,
+	// we know that the IBC transfer succeeded. That entails
+	// that the packet data is valid and can be safely
+	// deserialized without checking errors.
+	if ack.Success() {
+		// extract the coin info received from the packet data
+		var data ibctransfertypes.FungibleTokenPacketData
+		_ = types.ModuleCdc.UnmarshalJSON(packet.GetData(), &data)
+
+		// check if the recipient is the consumer reward's pool address
+		receiver, _ := sdk.AccAddressFromBech32(data.Receiver)
+		if receiver.String() != im.keeper.GetConsumerRewardsPoolAddressStr(ctx) {
+			return ack
```

Comment:
I'm curious about these `return ack`s- how do they work in the context of middleware?

I'm assuming that when you do this is just passes control flow out of the middleware, so if the packet is going somewhere else it can?

Reply:
These `ack`s are returned directly to [the IBC core RecvPacket handler](https://github.com/cosmos/ibc-go/blob/v7.3.1/modules/core/keeper/msg_server.go#L453), since in our context, we've only one middleware wrapping the transfer module.